feat(linux): linux support

Enubia opened the pull request:
Missing things:

- [x] testing ingame on linux

Enubia commented:
Missing things:

- [ ] skip/fallback for first linux release version check (throws error while checking for updates)
- [ ] testing ingame on linux

Enubia commented:
Resolved with https://github.com/PoE-Overlay-Community/PoE-Overlay-Community-Fork/pull/18

Enubia commented:
reopened because of dev setup documentation

Enubia commented:
@ijemafe @wratho @kifers whoever reviews this, the important files are game.ts, main.ts and window.ts, those contain the changes needed to run it on linux

Enubia commented on main.ts:
```diff
@@ -1,339 +1,377 @@
+      } else {
+        win.setAlwaysOnTop(true)
+        win.maximize()
+      }
```
this is the hack that loads the window, apparently windows do not get maximized with setAlwaysOnTop only so I had to do that manually

ijemafe replied:
Making sure I understand what `win.maximize()` does - if someone is playing in windowed mode, but not fullscreen, will this resize their game window?

Enubia replied:
```win.maximize()``` just maximizes the poe overlay clickthrough window on which the remaining application is drawn on, we don't have access to the ```level``` parameter on linux and because of that the window would start minimized. Since that's in conflict with the remainder of the overlay windows, I had to do that manually. This won't interact with the game in any way. The ```level``` parameter does something similar on windows, it pulls the overlay window into forground and thus it doesn't get minimized.

ijemafe replied:
Right, thanks for the explanation.

Enubia commented on electron-builder.json:
```diff
@@ -30,7 +30,7 @@
     "linux": {
-        "icon": "dist/favicon@4x.png",
+        "icon": "dist/favicon256x256.png",
         "target": [
```
on linux, icons need to have their size explicitly contained in their name

thorsten-passfeld commented:
I've already been using POE-Overlay on Linux in the past while helping the creator test it sometimes. I think I remember that we talked about node-window-manager maybe having to be replaced in the future, because it's missing some important features on Linux or because some features are missing on Linux.
What I always found annoying is that you had to alt-tab with nothing behind the POE window to be able to click the popup. That was because the overlay didn't correctly register/grab the POE window and therefore couldn't "always be on top". If I remember correctly, there is a file responsible for this that contains a code section, which looks for the Window name or something.

Enubia commented:
@thorsten-passfeld 
I've replaced node-window-manager with another package if it detects a linux environment which should fix the on top problem. One feature from node-window-manager is the ```bringToTop()``` function which is not implemented in the new package but that shouldn't make to much of a difference because it is only there to focus the game once poe overlay closes it's search window. I've already checked the implementation from awakened-poe-trade because I wanted to know how they did it but they implemented all of the node-window-manager features them self I assume (the important parts are covered from the new package). The real problem why it wasn't staying in foreground is that on linux it didn't create the fullscreen window when you are ingame which I fixed by setting ```win.maximize()``` if linux is detected. The real problem is that there is no level restriction for linux windows. This sets on which layer the window interacts. For Windows and mac it's set to ```'pop-up-menu'``` which is higher than normal windows. On linux we can't do that because it's simply not supported. Hence the sometimes buggy interaction with poe-overlay. For the most part it works, tested it on Gnome with Ubuntu 20.04 running in Wine with Vulcan enabled.

ijemafe commented on electron/window.ts:
```diff
@@ -1,34 +1,63 @@
-}
\ No newline at end of file
+  } catch (error) {
+    console.warn('Could not get active window.', error)
```
Should this be more than a warning?

Enubia replied:
Maybe, although this shouldn't happen

Enubia replied:
getActiveWindow() is called on every game update cycle, so if one time the window is undefined it shouldn't matter to much I'd say.

ijemafe commented on electron/game.ts:
```diff
@@ -1,85 +1,89 @@
-    public focus(): void {
-        this.window?.bringToTop();
+  public focus(): void {
+    if (this.window?.bringToTop) {
```
Is the `?` here checking for `undefined` values from `getActiveWindow()`? Are undefined windows fatal to the Overlay?

Enubia replied:
Yes exactly, if window would be undefined it would throw an exception on the bringToTop function which is obviously not available on undefined

ijemafe commented on main.ts:
```diff
@@ -1,339 +1,377 @@
+    win.setAlwaysOnTop(true, 'pop-up-menu', 1)
+  } else {
+    win.setAlwaysOnTop(true)
+  }
```
I see this code repeats from line 135 above. Should this be pulled out into a helper function that's called from each place?

Enubia replied:
I don't think so

The first initial call to this is on app init and the other specific one on game update which sets different parameters
